src/Workspaces/Core/Portable/Remote/IRemoteKeepAliveService.cs

```diff
@@ -18,31 +20,157 @@ internal interface IRemoteKeepAliveService
+                   (service, solutionInfo, cancellationToken) => service.KeepAliveAsync(solutionInfo, sessionId, cancellationToken),
+                   session.KeepAliveTokenSource.Token).ConfigureAwait(false);
+            }
+            catch (Exception ex) when (FatalError.ReportAndCatchUnlessCanceled(ex))
+            {
+                // If *anything* (other than cancellation) goes wrong here, we need to make sure we dispose the session
+                // fully.  Otherwise, we risk the call to WaitForSessionIdAsync making its way to OOP and never
+                // returning.
+                //
+                // Note: this would happen in a catastrophic failure scenario (like something going totally wrong with
+                // our host/oop connection). However, we don't want to exacerbate things by causing whatever is creating
+                // the keep alive session to entirely hang waiting. This is why we always report an error here as it
+                // indicates something has gone very wrong.
+                //
+                // This works because Dispose here will cancel the KeepAliveTokenSource, which is mixed into the call to
+                // WaitForSessionIdAsync as well.
+                //
+                // It is *fine* to get a cancellation exception here.  That's the *expected* scenario when we properly
+                // return this KeepAliveSession to the caller, who then cleanly Disposes of us.  This will cancel
+                // 'KeepAliveTokenSource', which is exactly what will cause the KeepAliveAsync to actually return.
+                session.Dispose();
+
+                // Note: we are intentionally not bubbling this exception out.  This was a fire-and-forget call.  So the
+                // result is entirely unobserved.  We will have already reported non-cancellation errors in the
+                // FatalError.ReportAndCatchUnlessCanceled call above.  There's nothing more to be done here.
+            }
+        }
+
+        static async Task WaitForSessionIdAsync(
+            SolutionCompilationState compilationState,
+            ProjectId? projectId,
+            RemoteHostClient client,
+            RemoteKeepAliveSession session,
+            CancellationToken callerCancellationToken)
+        {
+            try
+            {
+                // Subtle, but critical for correctness.  We make a linked token that combines the caller's cancellation
+                // token and the session's KeepAliveTokenSource.Token.  This way, if either the caller cancels (they don't
+                // need the session to be established), or if the session itself is disposed (which can happen if we fail
+                // inside InvokeKeepAliveAsync), we can bail out of this WaitForSessionIdAsync.
+                using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(
+                    session.KeepAliveTokenSource.Token,
+                    callerCancellationToken);
```

Comment:
this also changed.  We want to make sure that cleanup (which can now happen from KeepAliveAsync) ensures that WaitForSessionIdAsync cleans up too.  Mixing the two cancellation tokens together ensures that happens.